### ISSUE TITLE
Design database

### ISSUE THREAD

**Issue by McSinyx** [Huy-Ngo/gpms#2]
IIRC per our offline conversation, we reached a consensus to not use a relational database.  Instead, we're going to use a file-system-like database, where:

- Concurrency is managed by the caller
- IPFS is used for revision and rollback, or to be more precised,
  - The so-called file system is managed inside a MFS
  - For each modification, we log the hash of the data's root directory to a file inside of MFS and pin it to avoid the old revisions from being garbage-collected
  - Optionally unpin old revisions to GC them and save disk space

Basically, what IPFS MFS does is quite similar to git, but

1. It works better with larger files
2. It's more scalable (via ipfs clusters)

Compare to a RDBMS, IPFS also makes handling actual files (which will be quite many, i.e. as proof of works to finish tasks) less of a pain in the ass.

These technical choices shouldn't affect how the data structure is designed though.

**Comment by Huy-Ngo** [comment]
I'll help with this due to the time constraint. I should put the database design in docs/solution right?
